torchao/quantization/quantize_/int4_groupwise_preshuffle_tensor.py

```diff
@@ -0,0 +1,343 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD 3-Clause license found in the
+# LICENSE file in the root directory of this source tree.
+
+
+import importlib.util
+from typing import List
+
+import torch
+from torch.utils._python_dispatch import return_and_correct_aliasing
+
+from torchao.utils import (
+    TORCH_VERSION_AT_LEAST_2_5,
+    TorchAOBaseTensor,
+    fill_defaults,
+)
+
+__all__ = [
+    "Int4GroupwisePreshuffleTensor",
+]
+
+aten = torch.ops.aten
+
+
+if importlib.util.find_spec("fbgemm_gpu") is None:
+    quantize_int4_preshuffle = None
+else:
+    from fbgemm_gpu.experimental.gen_ai.quantize import quantize_int4_preshuffle
+
+
+class Int4GroupwisePreshuffleTensor(TorchAOBaseTensor):
+    """
+    Groupwise int4 weight only quantization
+
+    Tensor Attributes:
+        packed_weight: packed int4 weight, either 2D (N, K/2) or 3D (B, N, K/2), last dimension is packed
+        group_scale: (K/group_size, N) for 2D Tensor, (B, N, K/group_size) for 3D Tensor
+               dtype is the same as the original Tensor dtype
+        group_zero: (K/group_size, N) for 2D Tensor, (B, N, K/group_size) for 3D Tensor
+               dtype is the same as the original Tensor dtype
+
+    Non-Tensor Attributes:
+        group_size: the group size for groupwise quantization
+        shape_multiplier: is the multipler from packed_weight to the real weight, since
+        we pack the weight for int4, for example, when we pack the last dimension for
+        a 2D tensor, the shape_multiplier will be [1, 2]
+        shape: shape of the original Tensor
+
+    Note:
+       preshuffle means the weight is rearranged for more efficient use of loading instructions
```

Comment:
it would be good to share the specifics of the preshuffle transformation, either here or via a link a user can follow